c/tskit/tables.c

```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+    tsk_id_t *node_map = NULL;
+    tsk_id_t *individual_map = NULL;
+    tsk_id_t *population_map = NULL;
+    tsk_id_t *site_map = NULL;
+    tsk_id_t *mutation_map = NULL;
+    tsk_table_collection_t tables;
+
+    ret = tsk_table_collection_copy(self, &tables, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_clear(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+    node_map = malloc(tables.nodes.num_rows * sizeof(*node_map));
+    individual_map = malloc(tables.individuals.num_rows * sizeof(*individual_map));
+    population_map = malloc(tables.populations.num_rows * sizeof(*population_map));
+    site_map = malloc(tables.sites.num_rows * sizeof(*site_map));
+    mutation_map = malloc(tables.mutations.num_rows * sizeof(*mutation_map));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, tables.nodes.num_rows * sizeof(*node_map));
+    memset(individual_map, 0xff, tables.individuals.num_rows * sizeof(*individual_map));
+    memset(population_map, 0xff, tables.populations.num_rows * sizeof(*population_map));
+    memset(site_map, 0xff, tables.sites.num_rows * sizeof(*site_map));
+    memset(mutation_map, 0xff, tables.mutations.num_rows * sizeof(*mutation_map));
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        nid = nodes[k];
+        ret = tsk_node_table_get_row(&tables.nodes, nid, &node);
+        if (ret < 0) {
+            goto out;
+        }
+        iid = tables.nodes.individual[nid];
+        if (iid != TSK_NULL) {
+            if (individual_map[iid] == TSK_NULL) {
+                tsk_individual_table_get_row(&tables.individuals, iid, &ind);
+                ret = tsk_individual_table_add_row(&self->individuals, ind.flags,
+                    ind.location, ind.location_length, ind.metadata,
+                    ind.metadata_length);
+                if (ret < 0) {
+                    goto out;
+                }
+                individual_map[iid] = ret;
+            }
+            iid = individual_map[iid];
+        }
+        pid = tables.nodes.population[nid];
+        if (pid != TSK_NULL) {
+            if (population_map[pid] == TSK_NULL) {
+                tsk_population_table_get_row(&tables.populations, pid, &pop);
+                ret = tsk_population_table_add_row(
+                    &self->populations, pop.metadata, pop.metadata_length);
+                if (ret < 0) {
+                    goto out;
+                }
+                population_map[pid] = ret;
+            }
+            pid = population_map[pid];
+        }
+        ret = tsk_node_table_add_row(&self->nodes, node.flags, node.time, pid, iid,
+            node.metadata, node.metadata_length);
+        if (ret < 0) {
+            goto out;
+        }
+        if (ret != k) {
```

Comment:
Why would ret not be equal to k? If this is checking for the presence of a bug it should be an assert (but, I think it's unnecessary?, could just assert this at the end of the loop once?

Reply:
> could just assert this at the end of the loop once?

good call, yes